### ISSUE TITLE
[RFC] Code RL Recipe: GRPO Training with Execution-Based Rewards

### ISSUE THREAD

**Issue by dhruvnigam93** [verl-project/verl#5531]
# Motivation

verl has excellent math/reasoning recipes (`r1`, `open_math_reasoning`, `dapo`, `entropy`, `prime`, etc.) but **no code generation recipe**, despite this being explicitly requested on the roadmap:

- [#708](https://github.com/verl-project/verl/issues/708): *"Build a baseline code recipe in verl main repo, using small models such as llama or qwen-7b"* and *"Curate dataset for code RL training (start with existing open source ones)"* — both unchecked
- [#710](https://github.com/verl-project/verl/issues/710): LiveCodeBench, codecontests, TACO listed as unchecked benchmark targets

Code generation is well-suited for RL because it has a uniquely clean reward signal — the generated solution either passes the test cases or it doesn't. No separate reward model is needed. I'd like to contribute a self-contained recipe to `verl-recipe` that fills this gap.

# Design

## Algorithm

**GRPO** — already well-supported in verl, no new algorithm implementation needed.

## Model

**Qwen2.5-7B-Instruct** as the default policy model, with **LoRA** for memory efficiency. This allows training on 4x A100 40GB (or fewer). The recipe will be model-agnostic — any HuggingFace causal LM should work by changing the config.

A **Qwen2.5-3B** full-weight config will also be provided as a lightweight alternative for users with limited hardware.

## Dataset

- **Training**: [APPS](https://huggingface.co/datasets/codeparrot/apps) — 5,000 train / 5,000 test competitive programming problems with test cases, diverse difficulty levels (introductory / interview / competition)
- **Evaluation**: [MBPP](https://huggingface.co/datasets/google-research-datasets/mbpp) (sanitized split) and [HumanEval](https://huggingface.co/datasets/openai/openai_humaneval) for pass@k metrics

The data preprocessing script will convert problems into verl's dataset format.

## Reward Function

Execution-based, computed via verl's existing [SandboxFusion integration](https://github.com/verl-project/verl/tree/main/verl/utils/reward_score/sandbox_fusion):

| Signal | Weight | Description |
|--------|--------|-------------|
| Test pass rate | 0.7 | Fraction of test cases passed (0.0 to 1.0) |
| Format compliance | 0.2 | Code is parseable Python, within length limit |
| Compilation | 0.1 | Code compiles without syntax errors |

All execution happens inside SandboxFusion containers — no unsafe code on the host.

## verl Components Reused (no modifications needed)

| Component | Usage |
|-----------|-------|
| `RayPPOTrainer` | Main training loop (GRPO mode) |
| `ActorRolloutRefWorker` | Standard FSDP actor + vLLM rollout |
| `SandboxFusion` | Safe code execution for reward computation |
| `RLDataset` | Dataset loading with custom preprocessing |

## Custom Components (recipe-local only)

| Component | Description |
|-----------|-------------|
| `code_reward.py` | Extract code from model output, submit to SandboxFusion, compute pass rate |
| `data_process.py` | Convert APPS/MBPP/HumanEval to verl dataset format |
| `main_eval.py` | Evaluation script computing pass@1 and pass@10 |

# Overall Structure

```
code_rl/
    README.md
    __init__.py
    config/
        code_rl_grpo.yaml
    data_process.py
    code_reward.py
    main_eval.py
    run_qwen2_5_7b_code_grpo.sh
    run_qwen2_5_3b_code_grpo.sh
```

# Plan

- [ ] Data preprocessing: download APPS, convert to verl format
- [ ] Reward function: code extraction + SandboxFusion execution + scoring
- [ ] Training config: GRPO config tuned for code (longer response length, appropriate sampling temperature)
- [ ] Training scripts: shell scripts for single-node multi-GPU training
- [ ] Evaluation: pass@1 and pass@10 on MBPP-sanitized and HumanEval
- [ ] Documentation: README with setup, hardware requirements, expected metrics
- [ ] PR to `verl-project/verl-recipe`

# Hardware Requirements

- Training: 4x A100 40GB (single node) with LoRA, or 8x for full-weight
- SandboxFusion: CPU container (can run on the same node)

# Open Questions

1. **Dataset choice** — APPS is the default, but should I also include LiveCodeBench or codecontests as alternative configs? Happy to add these as additional run scripts.
2. **SandboxFusion vs lightweight sandboxing** — SandboxFusion is the official integration, but some users may want a simpler Docker-based execution setup. Should the recipe support both?
3. **Response length** — Code solutions can be long. Planning to use 2048 max response tokens. Is this aligned with verl's async rollout performance for longer sequences?

# CC

@eric-haibin-lin @vermouth1992 @tongyx361

**Comment by dhruvnigam93** [comment]
I'm going to try and create the recepie.

**Comment by Kirrito-k423** [comment]
Excellent RFC! Code generation is indeed a natural fit for RL with execution-based rewards.

A few suggestions:

1. **Dataset alternatives**: LiveCodeBench would be valuable as it has temporal coverage (problems from different months) and avoids contamination issues. Consider adding it as an alternative config.

2. **Response length**: 2048 tokens may be tight for complex problems. APPS competition-level problems often have solutions exceeding this. Consider 4096 or implementing adaptive truncation based on problem difficulty.

3. **Reward shaping**: Your weighted reward is reasonable. One addition: consider penalizing repeated code patterns to encourage diversity in exploration.

4. **SandboxFusion vs Docker**: SandboxFusion is the right choice for the main recipe. A Docker-based fallback could be documented for users who can't run the full SandboxFusion setup.

5. **Evaluation timing**: Since code execution is fast, consider online evaluation during training (sample + evaluate every N steps) rather than only post-training evaluation.

6. **LoRA vs full fine-tuning**: For code, full fine-tuning may give better results since code patterns differ significantly from general instruction data. Worth benchmarking both.

Looking forward to the PR!
